rspace/src/main/scala/coop/rchain/rspace/history/instances/CachingHashHistoryReaderImpl.scala

```diff
@@ -0,0 +1,127 @@
+          case Some(DataLeaf(bytes)) =>
+            Sync[F].delay(
+              decodeDataWithIntermediaries[A](bytes).map(v => RichDatum(v._1, v._2))
+            )
+          case Some(p) =>
+            Sync[F].raiseError[Seq[RichDatum[A]]](
+              new RuntimeException(
+                s"Found unexpected leaf while looking for data at key $hash, data: $p"
+              )
+            )
+          case None => Seq.empty[RichDatum[A]].pure
+        }
+      )
+
+  /**
+    * read methods for continuations
+    */
+  override def getContinuations(hash: Blake2b256Hash): F[Seq[WaitingContinuation[P, K]]] =
+    getRichContinuations(hash).map(_.map(_.decoded))
+
+  // This methods returning raw bytes along with decode value is performnce optimisatoin
+  // Making diff for two Seq[(WaitingContinuation[P, K]] is 5-10 tims slower then Seq[ByteVector],
+  // so the second val of the tuple is exposed to compare values
+  override def getRichContinuations(
+      hash: Blake2b256Hash
+  ): F[Seq[RichWK[P, K]]] =
+    contsCache
+      .get(
+        HistoryPointer(h.root, hash),
+        fetchData(hash).flatMap {
+          case Some(ContinuationsLeaf(bytes)) =>
+            Sync[F].delay(
+              decodeContinuationsWithIntermediaries[P, K](bytes).map(v => RichWK(v._1, v._2))
+            )
+          case Some(p) =>
+            Sync[F].raiseError[Seq[RichWK[P, K]]](
+              new RuntimeException(
+                s"Found unexpected leaf while looking for continuations at key $hash, data: $p"
+              )
+            )
+          case None => Seq.empty[RichWK[P, K]].pure
+        }
+      )
+
+  /**
+    * read methods for joins
+    */
+  override def getJoins(hash: Blake2b256Hash): F[Seq[Seq[C]]] =
+    getRichJoins(hash).map(_.map(_.decoded))
+
+  override def getRichJoins(hash: Blake2b256Hash): F[Seq[RichJoin[C]]] =
+    joinsCache
+      .get(
+        HistoryPointer(h.root, hash),
+        fetchData(hash).flatMap {
+          case Some(JoinsLeaf(bytes)) =>
+            Sync[F].delay(decodeJoinsWithIntermediaries[C](bytes).map(v => RichJoin(v._1, v._2)))
+          case Some(p) =>
+            Sync[F].raiseError[Seq[RichJoin[C]]](
+              new RuntimeException(
+                s"Found unexpected leaf while looking for join at key $hash, data: $p"
+              )
+            )
+          case None => Seq.empty[RichJoin[C]].pure
+        }
+      )
+
+  /**
+    * Fetch data on a hash pointer
+    */
+  def fetchData(
+      key: Blake2b256Hash
+  ): F[Option[PersistedData]] =
+    h.find(key.bytes.toSeq.toList).flatMap {
+      case (trie, _) =>
+        trie match {
+          case LeafPointer(dataHash) => leafStore.get(dataHash)
+          case EmptyPointer          => Applicative[F].pure(None)
+          case _ =>
+            Sync[F].raiseError(new RuntimeException(s"unexpected data at key $key, data: $trie"))
+
+        }
+    }
+
+  override def root: Blake2b256Hash = h.root
```

Comment:
Let's have a rule for naming with single letter only in a _small_ scope to make the code more readable.

```scala
// OK
xs.map(x => x.doSomething)

// Not OK for class parameters (usually large scope)
class Foo(x: Bar)  {
  def doFooThing = x.doSomething
}

// Not OK for named functions
def foo(x: Bar) = {
  ... a lot of code
  val y = x.doSomething
}
```